src/vs/workbench/contrib/terminal/browser/terminalEditorService.ts

```diff
@@ -12,16 +13,52 @@ import { IEditorService } from 'vs/workbench/services/editor/common/editorServic
 export class TerminalEditorService extends Disposable implements ITerminalEditorService {
 	declare _serviceBrand: undefined;
 
-	terminalEditorInstances: ITerminalInstance[] = [];
+	instances: ITerminalInstance[] = [];
+	private _activeInstanceIndex: number = -1;
 
 	private _editorInputs: Map</*instanceId*/number, TerminalEditorInput> = new Map();
+	private _instanceDisposables: Map</*instanceId*/number, IDisposable[]> = new Map();
+
+	private readonly _onDidDisposeInstance = new Emitter<ITerminalInstance>();
+	get onDidDisposeInstance(): Event<ITerminalInstance> { return this._onDidDisposeInstance.event; }
+	private readonly _onDidChangeActiveInstance = new Emitter<ITerminalInstance | undefined>();
+	get onDidChangeActiveInstance(): Event<ITerminalInstance | undefined> { return this._onDidChangeActiveInstance.event; }
+	private readonly _onDidChangeInstances = new Emitter<void>();
+	get onDidChangeInstances(): Event<void> { return this._onDidChangeInstances.event; }
 
 	constructor(
 		@IEditorService private readonly _editorService: IEditorService
 	) {
 		super();
 
-		// TODO: Multiplex instance events
+		this._register(toDisposable(() => {
+			for (const d of this._instanceDisposables.values()) {
+				dispose(d);
+			}
+		}));
+		this._register(this._editorService.onDidActiveEditorChange(() => {
+			const activeEditor = this._editorService.activeEditor;
+			this.activeInstance = activeEditor instanceof TerminalEditorInput ? activeEditor?.terminalInstance : undefined;
+		}));
+	}
+
+	get activeInstance(): ITerminalInstance | undefined {
+		if (this.instances.length === 0 || this._activeInstanceIndex === -1) {
+			return undefined;
+		}
+		return this.instances[this._activeInstanceIndex];
+	}
+
+	set activeInstance(value: ITerminalInstance | undefined) {
+		const oldActiveIndex = this._activeInstanceIndex;
+		if (value === undefined) {
+			this._activeInstanceIndex = -1;
+		} else {
+			this._activeInstanceIndex = this.instances.findIndex(e => e === value);
+		}
+		if (oldActiveIndex !== this._activeInstanceIndex) {
```

Comment:
I tried this once within `terminalService's ``setActiveInstance `and it broke something sort of obscure. We can keep our eye out for this, maybe it'll be fine.

Reply:
Oh yeah I remember the issue. If I create two terminals, first activeInstanceIndex will be 1 then activeInstanceIndex will be 2. Then if I right click on the tab associated with #1 (which focuses, but doesn't set that as the active instance) and kill it, activeInstanceIndex won't have changed, so onDidChangeActiveInstance won't be fired and the tabs list won't be updated.

Reply:
Good point, I just spent a while figuring out the same but in the group service 🤦